### PR TITLE
Implement `is (not) defined` feature

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes https://github.com/rinja-rs/rinja/issues/62.

I'm not super happy about how I implemented the parser part (the little "cheat") but found it easier to be done this way to keep the same logic as the other operators. But it forced to have some more error checking.

The other downside is that it makes the code to generate the `if`/`else` branches a bit messy. If you have ideas on how to make it look better by any chance?

### CODE REVIEW

**Comment by Kijewski** [comment]
Sorry, but I won't find the time to review before Friday/Saturday.

File: rinja_parser/src/expr.rs [hunk 1]
```diff
@@ -188,7 +190,59 @@ impl<'a> Expr<'a> {
...
-    expr_prec_layer!(muldivmod, filtered, alt((tag("*"), tag("/"), tag("%"))));
+    expr_prec_layer!(muldivmod, is_defined, alt((tag("*"), tag("/"), tag("%"))));
+
+    fn is_defined(i: &'a str, level: Level) -> ParseResult<'a, WithSpan<'a, Self>> {
+        let (_, level) = level.nest(i)?;
+        let start = i;
+        let (i, left) = Self::filtered(i, level)?;
+        // This is cheating: to make it easier, we expect `defined` to be provided afterward
+        // in order to keep the same code logic as the other calls.
+        let (i, right) = many0(pair(tuple((ws(tag("is ")), ws(opt(tag("not "))))), |i| {
```
Comment (Kijewski) [inline rinja_parser/src/expr.rs:201]:
```rs
let (i, is_neg) = preceded(
    ws(keyword("is")),
    cut(terminated(
        opt(keyword("not")),
        ws(keyword("defined")),
    )),
)(i)?;
```

**Comment by GuillaumeGomez** [comment]
> Sorry, but I won't find the time to review before Friday/Saturday.

You say that and then provide suggestion right away. 😆 

But in any case, it's not in a hurry. So please review when you have time and motivation to keep this project fun for you. :)

**Comment by GuillaumeGomez** [comment]
Finally applied your suggestion for the parser. Looks much better like this, thanks!

**Comment by GuillaumeGomez** [comment]
> The other downside is that it makes the code to generate the `if`/`else` branches a bit messy. If you have ideas on how to make it look better by any chance?

Thinking some more about this, I think we can simply create a new very small AST for conditions which would simply remove the conditions (and their associated block) that won't be rendered and then the function can simply loop through them instead of the whole state machine. Gonna do that.

**Comment by GuillaumeGomez** [comment]
Moved the logic out of the `write_if` method. Much closer to what it originally was.

**Comment by Kijewski** [comment]
> Moved the logic out of the `write_if` method. Much closer to what it originally was.

Thank you, much easier to read! I give you a proper review soon-ish.

File: rinja_parser/src/expr.rs [hunk 2]
```diff
@@ -188,7 +190,56 @@ impl<'a> Expr<'a> {
     expr_prec_layer!(band, shifts, token_bitand);
     expr_prec_layer!(shifts, addsub, alt((tag(">>"), tag("<<"))));
     expr_prec_layer!(addsub, muldivmod, alt((tag("+"), tag("-"))));
-    expr_prec_layer!(muldivmod, filtered, alt((tag("*"), tag("/"), tag("%"))));
+    expr_prec_layer!(muldivmod, is_defined, alt((tag("*"), tag("/"), tag("%"))));
+
+    fn is_defined(i: &'a str, level: Level) -> ParseResult<'a, WithSpan<'a, Self>> {
+        let (_, level) = level.nest(i)?;
```
Comment (Kijewski) [inline rinja_parser/src/expr.rs:196]: I think we can to without the nesting check in here. We only traverse to a lower level, not the same or a higher level.

File: rinja_parser/src/expr.rs [hunk 3]
```diff
@@ -188,7 +190,56 @@ impl<'a> Expr<'a> {
...
+                    "expected `defined` or `not defined` after `is`",
+                    // We use `start` to show the whole `var is` thing instead of the current token.
+                    start,
+                )));
+            }
+            Some(Some(None)) => false,
+            Some(Some(Some(_))) => true,
+        };
+        match *lhs {
+            Self::Var(_) => {}
```
Comment (Kijewski) [inline rinja_parser/src/expr.rs:216]: Please capture the variable name in here.

File: rinja_parser/src/expr.rs [hunk 4]
```diff
@@ -188,7 +190,56 @@ impl<'a> Expr<'a> {
...
+                    start,
+                )));
+            }
+        }
+
+        Ok((
+            i,
+            WithSpan::new(
+                if is_neg {
+                    Self::IsNotDefined(Box::new(lhs))
```
Comment (Kijewski) [inline rinja_parser/src/expr.rs:235]: The use the variable name in here. No need for boxing. This makes the code a lot more simple, I think.

**Comment by Kijewski** [comment]
Sorry for the iterative reviews! It's a bit hard for me to wrap my head around this feature.

File: rinja_parser/src/expr.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply rinja_parser/src/expr.rs:196]: Don't we need that for `Self::filtered(i, level)?` just below?

File: rinja_parser/src/expr.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply rinja_parser/src/expr.rs]: That was an incredible idea! Code got simplified by quite a lot! :D

File: rinja_parser/src/expr.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply rinja_parser/src/expr.rs]: Done!

**Comment by GuillaumeGomez** [comment]
> Sorry for the iterative reviews! It's a bit hard for me to wrap my head around this feature.

It's fine. Don't hesitate if you need some clarifications or any help to understand the code (which, if not easy to understand, should be more documented! So don't hesitate to tell me what's unclear so I can add code comments :smiley:)

File: rinja_parser/src/expr.rs (hunk 2 above)
Comment (Kijewski) [reply rinja_parser/src/expr.rs:196]: No, you can just pass the `level` as is. In all other cases we also only increment the recursion level on recursions. The code cannot parse an expression like `y is defined is defined`, so no nesting happens in here.

**Comment by Kijewski** [comment]
I pushed a commit to remove the line of unsafe code. I actually wanted to look if it was easily possible to remove it, and the best test was implementing it. :D

The code looks good, and after reading your comments properly, I understand it perfectly (I think). Great work!

File: rinja_parser/src/expr.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply rinja_parser/src/expr.rs]: Good point! Removed this nesting.

**Comment by GuillaumeGomez** [comment]
Fixed the merge conflict and applied your suggestion. :)

> I pushed a commit to remove the line of unsafe code. I actually wanted to look if it was easily possible to remove it, and the best test was implementing it. :D

That's my first implementation. I removed it to reduce the number of lifetimes. If you prefer with references, then I'm fine with it. :)

> The code looks good, and after reading your comments properly, I understand it perfectly (I think). Great work!

Awesome! \o/

**Review by Kijewski — DISMISSED** [review]
Thank you! Please revert the changes in the ui test nesting tests, then you can merge the PR.

**Comment by GuillaumeGomez** [comment]
Reverted wrong nested ui tests changes.

**Review by Kijewski — APPROVED** [review]
Thank you, well done!

**Pull request merged**